docs/user-guide/generic-providers.ipynb

```diff
@@ -403,6 +405,58 @@
    "source": [
     "graph.compute()"
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Constraining type vars for a specific pipeline\n",
+    "\n",
+    "Sometimes, listing constraints when defining a type variable is too restrictive.\n",
+    "We need to know all possible types that the type variable can take when defining it.\n",
+    "However, there is a workaround.\n",
+    "We can specify constraints when defining a *pipeline* instead of when defining a *type variable*.\n",
+    "\n",
+    "In this example, `T` itself is unconstrained.\n",
+    "But all occurrences of `T` in `pipeline` will be constrained to `int` and `float`."
```

Comment:
Should we call this a workaround? I think it makes sense to promote this as a "well" supported mechanism, just like having constraints on the typevar? Maybe it should be mentioned right from the start (where constraints are first mentioned above) that there are two cases?

Reply:
Done